### ISSUE TITLE
Reason behind  dividing GRUs' outputs by 2

### ISSUE THREAD

**Issue by tRosenflanz** [mp2893/retain#2]
Hello Edward,

Thank you for sharing the code for the Retain paper! I have been looking through it and found that in attention_step function of build_model you've divided the GRUs' outputs by 2. What is the reasoning behind this operation? Relevant code:
```
def attentionStep(att_timesteps):
		reverse_emb_t = temb[:att_timesteps][::-1]
		reverse_h_a = gru_layer(tparams, reverse_emb_t, 'a', alphaHiddenDimSize)[::-1] * 0.5
		reverse_h_b = gru_layer(tparams, reverse_emb_t, 'b', betaHiddenDimSize)[::-1] * 0.5 
```

**Comment by mp2893** [comment]
Hi tRosenflanz,

I think that comes from the time when I used to apply dropout to the GRUs' outputs. 
Now that I don't use dropout on the GRUs' outputs, I think dividing by two doesn't do much.
(I think I didn't clean that up because it didn't cause any problem)

**Comment by tRosenflanz** [comment]
Sounds good. Theoretically, I don't see why it would change the final prediction since the output layer should rebalance the predictions. However, dividing these values in half does have an effect on alpha outputs due to the combination of tanh and softmax activations (less important visits have higher value and more important ones have lower one) - in some ways that is akin to regularizing alpha layer.

**Comment by mp2893** [comment]
That is precisely correct.
Now that I think of it, I think I saw a long time ago a Theano neural network code by someone else, where this type of regularization (i.e. halving the embedding layer to reduce the variance) was applied. So I just left them in my code as well. 
(I am not too sure because, after all, it has been over a year since I actively wrote RETAIN)
